programs/validator-history/src/instructions/copy_cluster_info.rs

```diff
@@ -36,26 +39,64 @@ pub fn handle_copy_cluster_info(ctx: Context<CopyClusterInfo>) -> Result<()> {
 
     let epoch_start_timestamp = cast_epoch_start_timestamp(clock.epoch_start_timestamp);
 
-    // Sets the slot history for the previous epoch, since the current epoch is not yet complete.
-    if epoch > 0 {
-        cluster_history_account
-            .set_blocks(epoch - 1, blocks_in_epoch(epoch - 1, &slot_history)?)?;
-    }
-    cluster_history_account.set_blocks(epoch, blocks_in_epoch(epoch, &slot_history)?)?;
+    let epoch_schedule = EpochSchedule::get()?;
+
+    let slot_history = if epoch > 0 {
+        let slot_history_next_slot = slot_history.next_slot;
+        let start_slot = epoch_schedule.get_first_slot_in_epoch((epoch - 1).into());
+        let end_slot = epoch_schedule.get_last_slot_in_epoch((epoch - 1).into());
+        let (num_blocks, bitvec_inner) =
+            confirmed_blocks_in_epoch(start_slot, end_slot, *slot_history)?;
+        // Sets the slot history for the previous epoch, since the total number of blocks in the epoch is now final
+        cluster_history_account.set_blocks(epoch - 1, num_blocks)?;
+        // recreates SlotHistory with same heap memory chunk
+        Box::new(SlotHistory {
+            bits: bitvec_inner.into(),
+            next_slot: slot_history_next_slot,
+        })
+    } else {
+        slot_history
+    };
+
+    let start_slot = epoch_schedule.get_first_slot_in_epoch(epoch.into());
+    let end_slot = epoch_schedule.get_last_slot_in_epoch(epoch.into());
+    let (num_blocks, _) = confirmed_blocks_in_epoch(start_slot, end_slot, *slot_history)?;
+    cluster_history_account.set_blocks(epoch, num_blocks)?;
     cluster_history_account.set_epoch_start_timestamp(epoch, epoch_start_timestamp)?;
 
     cluster_history_account.cluster_history_last_update_slot = clock.slot;
 
     Ok(())
 }
 
-fn blocks_in_epoch(epoch: u16, slot_history: &SlotHistory) -> Result<u32> {
-    let epoch_schedule = EpochSchedule::get()?;
-    let start_slot = epoch_schedule.get_first_slot_in_epoch(epoch as u64);
-    let end_slot = epoch_schedule.get_last_slot_in_epoch(epoch as u64);
+const BITVEC_BLOCK_SIZE: u64 = 64;
 
+pub fn confirmed_blocks_in_epoch(
+    start_slot: u64,
+    end_slot: u64,
+    slot_history: SlotHistory,
```

Comment:
pass by ref

Reply:
I had this as `Box<SlotHistory>` but clippy complained, and it somehow works without copying the var.
Can't be &SlotHistory because the function takes ownership at the `into_boxed_slice` line
<img width="753" alt="Screenshot 2024-03-06 at 4 46 40 PM" src="https://github.com/jito-foundation/stakenet/assets/8891180/7fdfc10b-7f5c-4dfe-b10e-5f7620fccb0f">